packages/workspace/src/node/default-workspace-server.ts

```diff
@@ -150,22 +151,21 @@ export class DefaultWorkspaceServer implements WorkspaceServer, BackendApplicati
     }
 
     async getRecentWorkspaces(): Promise<string[]> {
-        const listUri: string[] = [];
         const data = await this.readRecentWorkspacePathsFromUserHome();
         if (data && data.recentRoots) {
-            data.recentRoots.forEach(element => {
-                if (element.length > 0) {
-                    if (this.workspaceStillExist(element)) {
-                        listUri.push(element);
-                    }
+            const allRootUris = await Promise.all(data.recentRoots.map(element => new Promise<string | undefined>(async resolve => {
+                if (element && await this.workspaceStillExist(element)) {
+                    resolve(element);
```

Comment:
Suggestion: This can be simplified.

```ts
const allRootUris = await Promise.all(data.recentRoots.map(async element => element && await this.workspaceStillExist(element) ? element : undefined));
```

Reply:
https://github.com/eclipse-theia/theia/pull/12798/commits/e42fd5de59fcdafeb2b7d0e64047849e8d35e6d8